ethereum/core/src/main/java/org/hyperledger/besu/ethereum/trie/diffbased/bonsai/storage/flat/BonsaiArchiveFlatDbStrategy.java

```diff
@@ -0,0 +1,404 @@
+                e ->
+                    new Pair<Bytes32, Bytes>(
+                        Bytes32.wrap(trimSuffix(e.toArrayUnsafe())),
+                        Bytes.of(
+                            storage.getNearestBefore(ACCOUNT_INFO_STATE, e).get().value().get())));
+    return stream;
+  }
+
+  @Override
+  protected Stream<Pair<Bytes32, Bytes>> storageToPairStream(
+      final SegmentedKeyValueStorage storage,
+      final Hash accountHash,
+      final Bytes startKeyHash,
+      final Function<Bytes, Bytes> valueMapper) {
+    return storage
+        .streamFromKey(
+            ACCOUNT_STORAGE_STORAGE,
+            calculateArchiveKeyNoContextMinSuffix(
+                calculateNaturalSlotKey(accountHash, Hash.wrap(Bytes32.wrap(startKeyHash)))))
+        .map(e -> Bytes.of(calculateArchiveKeyNoContextMaxSuffix(trimSuffix(e.getKey()))))
+        .takeWhile(pair -> pair.slice(0, Hash.SIZE).equals(accountHash))
+        .distinct()
+        .map(
+            key ->
+                new Pair<>(
+                    Bytes32.wrap(trimSuffix(key.slice(Hash.SIZE).toArrayUnsafe())),
+                    valueMapper.apply(
+                        Bytes.of(
+                                storage
+                                    .getNearestBefore(ACCOUNT_STORAGE_STORAGE, key)
+                                    .get()
+                                    .value()
+                                    .get())
+                            .trimLeadingZeros())));
+  }
+
+  @Override
+  protected Stream<Pair<Bytes32, Bytes>> storageToPairStream(
+      final SegmentedKeyValueStorage storage,
+      final Hash accountHash,
+      final Bytes startKeyHash,
+      final Bytes32 endKeyHash,
+      final Function<Bytes, Bytes> valueMapper) {
+    return storage
+        .streamFromKey(
+            ACCOUNT_STORAGE_STORAGE,
+            calculateArchiveKeyNoContextMinSuffix(
+                calculateNaturalSlotKey(accountHash, Hash.wrap(Bytes32.wrap(startKeyHash)))),
+            calculateArchiveKeyNoContextMaxSuffix(
+                calculateNaturalSlotKey(accountHash, Hash.wrap(endKeyHash))))
+        .map(e -> Bytes.of(calculateArchiveKeyNoContextMaxSuffix(trimSuffix(e.getKey()))))
+        .takeWhile(pair -> pair.slice(0, Hash.SIZE).equals(accountHash))
+        .distinct()
+        .map(
+            key ->
+                new Pair<>(
+                    Bytes32.wrap(trimSuffix(key.slice(Hash.SIZE).toArrayUnsafe())),
+                    valueMapper.apply(
+                        Bytes.of(
+                                storage
+                                    .getNearestBefore(ACCOUNT_STORAGE_STORAGE, key)
+                                    .get()
+                                    .value()
+                                    .get())
+                            .trimLeadingZeros())));
+  }
+
+  /*
+   * Puts the account data for the given account hash and block context.
+   */
+  @Override
+  public void putFlatAccount(
```

Comment:
we cannot add direlctly the blocknumber in the method fields instead of passing a context ? like that the implementation decide ot use it or not ? just a proposition

Reply:
The latest refactoring is cleaner in this regard I believe.